src/bt.rs

```diff
@@ -91,310 +97,296 @@ impl<V> Node<V> {
+    /// Gets a reference to the node located at the end of the path
+    /// traversing the tree from the root node.
     ///
-    /// This function traverses the tree from this node (`self`) until reaching the
-    /// node at the end of the path. It returns [None], if the node is
-    /// unreachable or nonexistent. Otherwise, it returns a mutable reference
-    /// to the node.
-    pub fn get_node(&mut self, path: &Path) -> Option<&mut Node<V>> {
-        let mut node = self;
-        for bit in path {
-            match if !bit {
-                &mut node.left
-            } else {
-                &mut node.right
-            } {
-                None => return None,
-                Some(next_node) => node = next_node,
-            };
-        }
-        Some(node)
+    /// This function traverses the tree from the root node until reaching
+    /// the node at the end of the path. It returns [None], if the node is
+    /// unreachable or nonexistent; otherwise, it returns a reference to the
+    /// node.
+    pub fn get_node(&self, path: &Path) -> Option<NodeRef> {
+        self.get_node_at(NodeRef::ROOT, path)
     }
-}
-
-/// Represents an append-only binary tree.
-pub struct BinaryTree<V> {
-    root: SubTree<V>,
-}
 
-impl<V> BinaryTree<V> {
-    /// Inserts the value at the end of the path.
-    ///
-    /// This function traverses the tree from the root node until reaching the
-    /// node at the end of the path. If the node is unreachable or already
-    /// contains a value, an error wrapping the value is returned. Otherwise,
-    /// a new node containing the value is inserted.
-    ///
-    /// # Returns
-    /// - `Ok(())` when the node is inserted at the end of the path.
-    /// - `Err(InsertNonEmptyNode(value))` when the tree already contains a
-    /// value at the end of the path.
-    /// - `Err(UnreachablePath(value))` when the end of the path is unreachable.
+    /// Gets a reference to the node located at the end of the path
+    /// traversing the tree from the specified node reference.
     ///
-    /// In the error cases, no insertion occurs and the value is returned to
-    /// the caller of this function.
-    pub fn insert(&mut self, path: &Path, value: V) -> Result<(), BinaryTreeError<V>> {
-        if let Some(node) = &mut self.root {
-            node.insert(path, value)
-        } else if path.is_empty() {
-            self.root = Some(Box::new(Node::new(value)));
-            Ok(())
-        } else {
-            Err(BinaryTreeError::UnreachableNode(value))
+    /// This function traverses the tree from the specified node until
+    /// reaching the node at the end of the path. It returns [None],
+    /// if the node is unreachable or nonexistent; otherwise, it returns
+    /// a reference to the node.
+    pub fn get_node_at(&self, node_ref: NodeRef, path: &Path) -> Option<NodeRef> {
+        self.is_valid_node_ref(node_ref).then_some(())?;
+
+        let new_index = self.nodes.len();
```

Comment:
nit: `new_index` is unused?

Reply:
this is odd, why clippy didn't catch this issue

Reply:
This is suppressed by the module-level `#![allow(unused_variables)]`.